unixfs/src/dir/builder/iter.rs

```diff
@@ -0,0 +1,416 @@
+use super::{Entry, Leaf, TreeConstructionFailed, TreeOptions, Visited};
+use cid::Cid;
+use std::collections::{BTreeMap, HashMap};
+
+/// Constructs the directory nodes required for a tree.
+///
+/// Implements the Iterator interface for owned values and the borrowed version `next_borrowed`.
+/// Tree is fully constructed once this has been exhausted.
+pub struct PostOrderIterator<'a> {
+    pub(super) full_path: &'a mut String,
+    pub(super) old_depth: usize,
+    pub(super) block_buffer: &'a mut Vec<u8>,
+    // our stack of pending work
+    pub(super) pending: Vec<Visited>,
+    // "communication channel" from nested entries back to their parents
+    pub(super) persisted_cids: HashMap<Option<u64>, BTreeMap<String, Leaf>>,
+    pub(super) reused_children: Vec<Visited>,
+    pub(super) cid: Option<Cid>,
+    pub(super) total_size: u64,
+    // from TreeOptions
+    pub(super) wrap_with_directory: bool,
+}
+
+impl<'a> PostOrderIterator<'a> {
+    pub(super) fn new(
+        root: Visited,
+        full_path: &'a mut String,
+        block_buffer: &'a mut Vec<u8>,
+        opts: TreeOptions,
+    ) -> Self {
+        full_path.clear();
+
+        PostOrderIterator {
+            full_path,
+            old_depth: 0,
+            block_buffer,
+            pending: vec![root],
+            persisted_cids: Default::default(),
+            reused_children: Vec::new(),
+            cid: None,
+            total_size: 0,
+            wrap_with_directory: opts.wrap_with_directory,
```

Comment:
`old_depth`, `persisted_cids`, `reused_children`, `cid` and `total_size` can be initialized below with `..Default::default()`

Reply:
I was hopeful that I could just replace those lines to get copypaste from the rustc however using ..Default::default() requires implementing it in the first place.

Reply:
I thought it was derived; long reviews deceive the eyes 😃